### PR TITLE
Link: listen to history for active state changes

### PR DESCRIPTION
**Pull request by malte-wessel**
* subscribe to history changes
* don't use the context to pass data
* works with parents that implement shouldComponentUpdate
* solves #470, #1772

### CODE REVIEW

File: modules/Link.js [hunk 1]
```diff
@@ -54,6 +54,46 @@ var Link = React.createClass({
...
+    if (!this._unlisten) return;
+    this._unlisten();
+    this._unlisten = undefined;
+  },
+
+  handleHistoryChange() {
+    var { active } = this.state;
+    var nextActive = this.getActiveState();
+    if (active !== nextActive) {
+        this.setState({ active: nextActive });
```
Comment (gaearon) [inline modules/Link.js]: 4 spaces, should be 2

File: modules/__tests__/Link-test.js [hunk 2]
```diff
@@ -225,6 +225,43 @@ describe('A <Link>', function () {
     });
   });
 
+  describe('when route changes', function() {
+    it('changes active state', function(done) {
+      var LinkWrapper = React.createClass({
+        render() {
```
Comment (gaearon) [inline modules/__tests__/Link-test.js]: How about we put `shouldComponentUpdate() { return false }` on it just to be sure it works?

**Comment by gaearon** [comment]
You are my hero :heart: 

@mjackson @ryanflorence Can we please get your eyes on this? I really really want this to get into 1.0.

**Comment by mjackson** [comment]
niiiiiiiice ;) I was hoping someone would take a stab at this. Thanks @malte-wessel ! I'll take a closer look tomorrow morning.

**Comment by malte-wessel** [comment]
@gaearon spaces are fixed, shouldComponentUpdate is in :thumbsup:

**Comment by mjackson** [comment]
Looks solid to me, @malte-wessel! Thanks for your thoughtful contribution here. It really takes some time to wrap your head around this codebase, and I really appreciate you taking the time to do it :)

**Comment by malte-wessel** [comment]
thank you guys!

**Comment by roundrobin** [comment]
Awesome thanks for much, now I can get rid of an entire store!

**Comment by mjackson** [comment]
Hmm ... turns out this approach causes another subtle bug. :/

In our tests, we now cannot assert active state on links. Here's what the flow looks like:

1. we get a new `location`
2. `<Router>` picks up the change and updates its state, causing a re-render
3. we assert that links have the correct active state (and fail)
4. `<Link>`s get the location change and update their active state

So, while I really like the fact that `<Link>`s now work as descendants of components that `return false` from `shouldComponentUpdate`, I'm afraid we're going to have to think of another solution for this particular problem. We have reverted these changes in 50372fe249cae70d768fcf19ead93423ed8175f0.

**Comment by gaearon** [comment]
@mjackson Can you expand a bit? I don't fully understand what's going on (particularly 2 and 3). Any way I can help?

**Comment by tajo** [comment]
Uh, I still don't now how to solve this annoying issue (working active Link inside a subtree with shouldComponentUpdate doing shallowCompare). I don't want to use any workarounds (like fiddling with this.context inside of shouldComponentUpdate or having my own UberLink component).

What about at least trying to get the history object from `this.props.history` if `this.props.context` does not exist? It's should be one-liner.

https://github.com/rackt/react-router/blob/master/modules/Link.js#L101

**Comment by taion** [comment]
You can follow along in the discussion on #470. For my money, though, I think you just shouldn't use `shallowCompare` on those subtrees. If you embed a `Link`, the parents of the `Link` are not really "pure".

**Pull request merged**